### ISSUE TITLE
[Bug] Memory leak: Got the response to undefined request 9009

### ISSUE THREAD

**Issue by cv65kr** [temporalio/sdk-php#289]
### Describe the bug

"worker_exec: SoftJobError:\n\tsync_worker_exec_payload: LogicException: Got the response to undefined request 9002 in /src/vendor/temporal/sdk/src/Internal/Transport/Client.php:57\nStack trace:\n#0 /src/vendor/temporal/sdk/src/WorkerFactory.php(389): Temporal\\Internal\\Transport\\Client->dispatch(Object(Temporal\\Worker\\Transport\\Command\\SuccessResponse))\n#1

Looks like there is still a memory leak.

### Minimal Reproduction

Hard to find potencial issue, because it was discovered on production for couple days without redeployment. I am using almost all options available from SDK starting from SubWorkflows across to Search attributes.

My configuration

.rr.yaml
```
temporal:
    address: ${TEMPORAL_CLI_ADDRESS}
    namespace: ${TEMPORAL_NAMESPACE}
    cache_size: 100
    activities:
        allocate_timeout: 3600s
        num_workers: 9
        supervisor:
            max_worker_memory: 164

```

It's scaled horizontally 10 times by default.

I don't know how I can help you to find issue.

### Environment/Versions

- OS and processor: Linux
- Temporal Version: 1.18.1
- SDK - 2.3.2
- AWS

**Comment by roxblnfk** [comment]
@cv65kr could you add RoadRunner version?

**Comment by wolfy-j** [comment]
Please share if you have any of:
1) Previous log outputs, did worker die?
2) Memory metrics ahead of the crash.

**Comment by cv65kr** [comment]
@roxblnfk @wolfy-j 
RR version 2.12.2

ECS task 1:
<img width="632" alt="image" src="https://user-images.githubusercontent.com/9404962/217220707-c975c321-458c-40f6-a60a-a0fda9e6f9d2.png">

ECS task 2:
<img width="632" alt="image" src="https://user-images.githubusercontent.com/9404962/217220902-90200de7-f695-49b3-af83-a70b2e09468b.png">

Unfortunately I don't have a metrics from RR itself, only those from AWS.

```
2023-02-07T07:32:44.535Z	ERROR	server      	no free workers in the pool, wait timeout exceed	{"reason": "no free workers", "internal_event_name": "EventNoFreeWorkers", "error": "worker_watcher_get_free_worker: NoFreeWorkers:\n\tcontext deadline exceeded"}

2023-02-07T07:32:44.847Z WARN server worker stopped, and will be restarted {"reason": "worker error", "pid": 640, "internal_event_name": "EventWorkerError", "error": "worker_exec: SoftJobError:\n\tsync_worker_exec_payload: LogicException: Got the response to undefined request 17079 in /src/vendor/temporal/sdk/src/Internal/Transport/Client.php:57\nStack trace:\n#0 /src/vendor/temporal/sdk/src/WorkerFactory.php(389): Temporal\\Internal\\Transport\\Client->dispatch(Object(Temporal\\Worker\\Transport\\Command\\SuccessResponse))\n#1 /src/vendor/temporal/sdk/src/WorkerFactory.php(261): Temporal\\WorkerFactory-

2023-02-07T07:32:44.847Z ERROR temporal Workflow panic {"Namespace": "default", "TaskQueue": "default", "WorkerID": "default:b450078f-8739-4167-a74a-628e35151e15", "WorkflowType": "myWorkflow", "WorkflowID": "1234444", "RunID": "e71e05fc-cefe-408e-8cc8-597966e96177", "Attempt": 1, "Error": "worker_exec: SoftJobError:\n\tsync_worker_exec_payload: LogicException: Got the response to undefined request 17079 in /src/vendor/temporal/sdk/src/Internal/Transport/Client.php:57\nStack trace:\n#0 /src/vendor/temporal/sdk/src/WorkerFactory.php(389): Temporal\\Internal\\Transport\\Client->dispatch(Object(Temporal\\Worker\\Transport\\Command\\SuccessResponse))\n#1 /src/vendor/temporal/sdk/src/WorkerFactory.php(261): 

2023-02-07T07:32:45.326Z WARN temporal Failed to poll for task. {"Namespace": "default", "TaskQueue": "default", "WorkerID": "default:b450078f-8739-4167-a74a-628e35151e15", "WorkerType": "WorkflowWorker", "Error": "worker stopping"
```


I extract unique logs, the rest it more likely `worker_exec: SoftJobError:\n\tsync_worker_exec_payload: LogicException: Got the response to undefined request 17029 in`

**Comment by cv65kr** [comment]
@roxblnfk @wolfy-j 

Container1:
```
Workers of [temporal]:
+---------+-----------+---------+---------+---------+--------------------+
|   PID   |  STATUS   |  EXECS  | MEMORY  |  CPU%   |      CREATED       |
+---------+-----------+---------+---------+---------+--------------------+
|     623 | ready     |   1,124 | 61 MB   |    0.03 | 1 hour ago         |
|     521 | ready     |      39 | 76 MB   |    0.02 | 1 hour ago         |
|     520 | ready     |      39 | 83 MB   |    0.02 | 1 hour ago         |
|     526 | ready     |      38 | 74 MB   |    0.02 | 1 hour ago         |
|     511 | ready     |      39 | 75 MB   |    0.02 | 1 hour ago         |
|     513 | ready     |      39 | 75 MB   |    0.02 | 1 hour ago         |
|     514 | ready     |      40 | 85 MB   |    0.02 | 1 hour ago         |
|     515 | ready     |      39 | 85 MB   |    0.03 | 1 hour ago         |
|     512 | ready     |      39 | 112 MB  |    0.04 | 1 hour ago         |
+---------+-----------+---------+---------+---------+--------------------+
```

Container 2:
```
Workers of [temporal]:
+---------+-----------+---------+---------+---------+--------------------+
|   PID   |  STATUS   |  EXECS  | MEMORY  |  CPU%   |      CREATED       |
+---------+-----------+---------+---------+---------+--------------------+
|     627 | ready     |     865 | 60 MB   |    0.02 | 1 hour ago         |
|     516 | ready     |      37 | 75 MB   |    0.02 | 1 hour ago         |
|     529 | ready     |      37 | 76 MB   |    0.02 | 1 hour ago         |
|     530 | ready     |      37 | 76 MB   |    0.02 | 1 hour ago         |
|     515 | ready     |      37 | 83 MB   |    0.02 | 1 hour ago         |
|     517 | ready     |      38 | 104 MB  |    0.03 | 1 hour ago         |
|     518 | ready     |      37 | 75 MB   |    0.02 | 1 hour ago         |
|     519 | ready     |      37 | 75 MB   |    0.02 | 1 hour ago         |
|     520 | ready     |      37 | 75 MB   |    0.02 | 1 hour ago         |
+---------+-----------+---------+---------+---------+--------------------+
```

1. One of worker uses more memory in compare to other processes
2. Why first process have such big amount of execs in compare to rest of processes?

**Comment by cv65kr** [comment]
@rustatian thanks of explanation of point 2.

I am quite sure that there is a leak somewhere 
<img width="615" alt="image" src="https://user-images.githubusercontent.com/9404962/217288988-f907516f-857e-427b-a245-cf6caef930a3.png">

I am trying to isolate workflow with separate worker to give you more details. I am not sure if this is a leak or not because here I have minimal traffic. In 3 hour it grow up with 3 mb, so it sounds like something gentle but if we imagine that multiply by some traffic it might it.
<img width="615" alt="image" src="https://user-images.githubusercontent.com/9404962/217289561-7a68e580-cbd7-427e-90fd-03882b4c5272.png">

**Comment by cv65kr** [comment]
View of workflow after `Got the response to undefined request 17029 in`. This is situation when workflow was not able to run subworkflow.
<img width="1569" alt="image" src="https://user-images.githubusercontent.com/9404962/217293667-76d1da75-67e3-4662-8804-7d50a5647191.png">

**Comment by rustatian** [comment]
@cv65kr If we would talk only about the workflow execution (without taking into account the restart reasons), this error will only lead to the workflow retrying, and as a result, your workflow will still finish correctly, am I right?

**Comment by cv65kr** [comment]
@rustatian unfortunately no, if you look on my comment with Temporal UI screen you will see that workflow was canceled (not by me). 

In the past I had a situation when because of this error couple sub-worklows was assigned to wrong parent and in result of that activity was executed with wrong input. It happened only one time and I was not able to reproduce it. 

That's why for me "Got the response to undefined request" error sounds very dangerous.

My first suspect was that horizontal scaling issues e.g. I sent signal to Temporal and Temporal notified all workers so in fact only worker with assigned workflow was able to clean in_memory cache after receiving the signa, and for all rests of workers stayed in memory.

**Comment by rustatian** [comment]
Could you let me know if you configured workflow retries?

**Comment by cv65kr** [comment]
I have workflows with unique id, so in general it cannot be retries. Only activities are retried in my case.

**Comment by cv65kr** [comment]
Hmm. No worries, but think about e.g. financial applications where something need to happen only one time and you need to have zero trust to avoid some double payouts.

**Comment by rustatian** [comment]
Mmmm, money 💵 
Are you 100% sure that your code doesn't have any leaks?

**Comment by cv65kr** [comment]
I am still checking, I cannot answer you for 100%. There is too much things to check. 
Right now I used RR soft reset just to be safe and still looking for root cause.

**Comment by rustatian** [comment]
Actually, `./rr reset` is not a soft one :) This command resets all pools of workers (WF+ACT), temporal workers and purges the sticky cache.

**Comment by cv65kr** [comment]
ttl under supervisor option as well?

**Comment by rustatian** [comment]
You can't set a supervisor on a WF worker, only activities. But for the activities, it resets only activities workers.

**Comment by rustatian** [comment]
WF worker contains all info about WF executions, responses, etc. It should not be restarted w/o a significant reason. You can see it first in the `./rr worker` output.

**Comment by cv65kr** [comment]
Yes. I mean for activities, and generally pass the exam (only for searching bug purposes)
<img width="1465" alt="image" src="https://user-images.githubusercontent.com/9404962/217853344-06190f3e-9143-4638-ba6b-09f8f8dda27f.png">
 
There was no cancelled/failed workflows/activities.

**Comment by wolfy-j** [comment]
Hi guys, 

could you please report if you still seeing this behaviour on recent SDK and RR releases? We included new mechanism which should improve worker restarts and prevent leaky ids in workers.

**Comment by cv65kr** [comment]
Sorry, but I am not able to test it now. I will try to do this with other project.

**Comment by wolfy-j** [comment]
To be finally fixed in 2.7 realase.

**Comment by roxblnfk** [comment]
Hello!
Could you try to use PHP SDK 2.6.0 (dev) with RoadRunner 2023.3.0-beta?

We have tested it on a lot of parallel workflows and signals with the `cache_size` option set to 1.
PHP SDK 2.6.0 with RoadRunner 2023.3.0-beta shows a perfect result.
No any "Got the response to undefined request" errors!

More details about PHP SDK 2.6.0 (dev) are there:
https://discord.com/channels/538114875570913290/824234752692977664/1142176919786360894
https://temporalio.slack.com/archives/C01LK9FAMM0/p1692386951356089

To install **SDK 2.6.0** dev and RoadRunner 2023.3.0-beta:

```bash
composer require temporal/sdk ~2.6.0@dev -W
vendor/bin/rr get --stability=beta
```

**Comment by roxblnfk** [comment]
Hello. Have you already updated to RoadRunner 2023.3 and PHP SDA 2.6.1? Is your problem reproducible on the new versions?

**Comment by wolfy-j** [comment]
Closing as solved until further comments.

**Comment by cv65kr** [comment]
@roxblnfk @wolfy-j @rustatian I was opportunity to test the solution during the load test and looks like it not working I see bunch of errors like:
```
"Error": "undefined response: Got the response to undefined request 11264", "StackTrace": "process event for test-qe [panic]:\ngithub.com/temporalio/roadrunner-temporal/v4/aggregatedpool.(*Workflow).OnWorkflowTaskStarted(0xc00525cf00, 0x14?)\n\tgithub.com/temporalio/roadrunner-temporal/v4@v4.5.6/aggregatedpool/workflow.go:172 +0x319\ngo.temporal.io/sdk/internal.(*workflowExecutionEventHandlerImpl).ProcessEvent(0xc003d7edf8, 0xc003c2d880, 0x80?, 0x1)\n\tgo.temporal.io/sdk@v1.25.1/internal/internal_event_handlers.go:1083 +0x229\ngo.temporal.io/sdk/internal.(*workflowExecutionContextImpl).ProcessWorkflowTask(0xc0008ecb40, 0xc001a95080)\n\tgo.temporal.io/sdk@v1.25.1/internal/internal_task_handlers.go:1074 +0x155d\ngo.temporal.io/sdk/internal.(*workflowTaskHandlerImpl).ProcessWorkflowTask(0xc002863ba0, 0xc001a95080, 0xc0008ecb40, 0xc0021bee40)\n\tgo.temporal.io/sdk@v1.25.1/internal/internal_task_handlers.go:868 +0x3bf\ngo.temporal.io/sdk/internal.(*workflowTaskPoller).processWorkflowTask(0xc0039e9e60, 0xc001a95080)\n\tgo.temporal.io/sdk@v1.25.1/internal/internal_task_pollers.go:354 +0x3a3\ngo.temporal.io/sdk/internal.(*workflowTaskPoller).ProcessTask(0xc0039e9e60, {0x1b414c0?, 0xc001a95080?})\n\tgo.temporal.io/sdk@v1.25.1/internal/internal_task_pollers.go:318 +0x87\ngo.temporal.io/sdk/internal.(*baseWorker).processTask(0xc003990dc0, {0x1b42040?, 0xc003ff33a0})\n\tgo.temporal.io/sdk@v1.25.1/internal/internal_worker_base.go:505 +0x15b\ngo.temporal.io/sdk/internal.(*baseWorker).processTaskAsync.func1()\n\tgo.temporal.io/sdk@v1.25.1/internal/internal_worker_base.go:356 +0x45\ncreated by go.temporal.io/sdk/internal.(*baseWorker).processTaskAsync in goroutine 20271\n\tgo.temporal.io/sdk@v1.25.1/internal/internal_worker_base.go:352 +0xb8"
```

**Comment by rustatian** [comment]
Hey @cv65kr 👋 
What is your RR and `sdk-php` versions?

**Comment by cv65kr** [comment]
@rustatian what I saw is also after redeployment there is an error `"Error": "nondeterministic workflow: extra replay command for CancelTimer: (TimerId:17)"}`

**Comment by wolfy-j** [comment]
Can you let us know which version did you update it from?

**Comment by cv65kr** [comment]
@wolfy-j 
RR 2.12.2 => 2023.3.6
SDK 2.4.0 => 2.6.1

But I ran totally new workflows

**Comment by wolfy-j** [comment]
Does this workflow use signals? There was a patch in 2.5.* to address some stability issues.

**Comment by cv65kr** [comment]
@wolfy-j yes,  I mentioned about that `worfklows are simple - some activity with http calls, await for signal with timeout`

> Does this workflow use signals? There was a patch in 2.5.* to address some stability issues.

Like you see I upgraded everything to the newest versions.

**Comment by wolfy-j** [comment]
Ok, thank you for clarifying. As for signal handling - there are some edge-case scenarios that were handled incorrectly prior to ~2.5 version. This has been addressed, but older workflows might still have this issue.

As for the undefined workflow response... this is odd, we've used the solution that eliminates the possibility of de-sync on a conceptual level. Valery is currently on vacations, give us some time to get back with more questions to you.

**Comment by cv65kr** [comment]
Feel free to raise questions, currently I have possibilities to do some load tests. I am attaching some metrics which might you reflects what's happened inside containers during running a 300 workflows batch in 5 sec
<img width="1331" alt="image" src="https://github.com/temporalio/sdk-php/assets/9404962/ef437551-b6ba-42db-8321-f509f4bf85bd">

So `"Error": "nondeterministic workflow: extra replay command for CancelTimer: (TimerId:17)"}` might related with `undefined request `?

**Comment by cv65kr** [comment]
I attaching couple metrics, which shows the moment when errors start pop up

<img width="1599" alt="image" src="https://github.com/temporalio/sdk-php/assets/9404962/862dc6e4-17c5-4013-a54a-a1ccaadbfba1">

CPU was high
<img width="176" alt="image" src="https://github.com/temporalio/sdk-php/assets/9404962/04bf9f34-1821-4fe3-bdb6-fcf93ec554eb">


Changed configuration for tests
```
->withMaxConcurrentWorkflowTaskPollers(
                    5
                )
                ->withMaxConcurrentActivityTaskPollers(
                    5
                )
                ->withMaxConcurrentActivityExecutionSize(
                    300
                )
                ->withMaxConcurrentWorkflowTaskExecutionSize(
                    300
                )
```

**Comment by wolfy-j** [comment]
> So `"Error": "nondeterministic workflow: extra replay command for CancelTimer: (TimerId:17)"}` might related with `undefined request `?

Yes, theoretically, it's possible that the older version of the workflow triggered the timer incorrectly (before the patch), and it might cause an extra callback. Do you have a chance to run these tests on post 2.5.3 workflows? I'm trying to ensure that de-sync issue is gone (it is based on our aggressive tests) and this problem is mostly related to post 2.5.3 behaviour.

**Comment by cv65kr** [comment]
@wolfy-j I tested now with 2.5.1 version and issue is the same.

**Comment by rustatian** [comment]
You should avoid `2.5.1` and test on `2.6.1` 😃

**Comment by cv65kr** [comment]
@rustatian I tested both and effect is the same :)

**Comment by rustatian** [comment]
I mean, `2.5.1` should not be involved in testing, since WF created with the outdated version might be triggered. Ideally, should be 0 WFs and testing session should be on the latest versions only, if that possible for sure.

**Comment by cv65kr** [comment]
Any updates?

**Comment by wolfy-j** [comment]
Hi, this is our pipeline to check at the beginning of Jan. We have to wrap current release first.

**Comment by wolfy-j** [comment]
Hi,

We've spent some time with the whole team over the last few days, and we can confirm that there are no memory leaks in the PHP codebase, including your use case. 

However, we managed to reproduce your issue quickly, and its root cause is pretty straightforward. So, let's go so we can finally close this ticket.

Issue number one - **"Task processing failed with error - Workflow task not found"**. This issue is NOT connected to a memory leak but is related to your configuration. 

In your current setup, it means that you can consume **300** workflows per worker in parallel - https://github.com/cv65kr/samples-php/blob/leak/app/worker.php#L37

Setting such value on a small worker is possible, but it will sometimes trigger workflow **task timeout** if the worker task can not be performed in time. The reason why it can't be performed in time:
1) Too much to unwind
2) Too many workflows per worker are trying to do something in parallel.

Next, you had two conflicting values in your workers:
1) cache size = 15000
2) PHP max_memory_limit = 128Mb (in this docker)

As a result, your worker appeared too hungry, trying to serve as many workflows from memory as possible until it hit OOM. Make sure that your sticky cache is set reasonably for your workflows. 

For your particular example, 500 workflows in memory consume about 140 MB, so a 15000 sticky cache will need around 4GB.

I've updated your code with:
1) cache size = 500
2) concurrent workflows = 200
3) workflow pollers = 15

![Screenshot 2024-01-16 153241](https://github.com/temporalio/sdk-php/assets/796136/e56760b0-3cbf-44fe-8745-472fab51ae6e)

As a result, we did not experience any failures or memory leaks in your code. In addition, we have found that under large workflow volume GC will slow down workflow worker, causing even more issues. 

Over the next few days, we will release a patch to address GC overhead, increasing the performance of use cases like yours by 3x.

![Screenshot 2024-01-16 161004](https://github.com/temporalio/sdk-php/assets/796136/e31deaad-631c-4e9a-aab5-7fac2d7adcda)

In short:
1) No memory leaks, your simply never saturated your worker
2) Please make sure to avoid using PHP memory limit in combination with dispoportional sticky cache size - worker killed by OOM causing ID sync issues
3) Do not overconsume in your workers

**Comment by wolfy-j** [comment]
Also, SDK version in your sample repo is 2.6.1.
